File: Time_Warp_Windows/TimeWarp.Wpf/Executors/PilotExecutor.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace TimeWarp.Wpf
{
    public class PilotExecutor
    {
        private readonly Dictionary<string, int> _labels = new(StringComparer.OrdinalIgnoreCase);
        private readonly Dictionary<string, string> _vars = new(StringComparer.OrdinalIgnoreCase);
        private bool _lastCondition;

        public string Execute(string code)
        {
            _labels.Clear(); _vars.Clear(); _lastCondition = false;
            var lines = new List<string>();
            foreach (var raw in code.Replace("\r", string.Empty).Split('\n'))
            {
                var t = raw.Trim(); if (t.Length > 0) lines.Add(t);
            }
            // First pass: labels L:NAME
            for (int i = 0; i < lines.Count; i++)
            {
                var up = lines[i].ToUpperInvariant();
                if (up.StartsWith("L:"))
                {
                    var name = lines[i].Substring(2).Trim();
                    if (name.Length > 0) _labels[name.ToUpperInvariant()] = i;
                }
            }

            int pc = 0; int steps = 0; var outBuf = new StringBuilder();
            while (pc < lines.Count && steps++ < 10000)
            {
                var line = lines[pc]; pc++;
                var up = line.ToUpperInvariant();
                if (up.StartsWith("L:")) continue;
                if (up.StartsWith("R:")) break;
                if (up.StartsWith("T:"))
                {
                    var text = line.Substring(2).Trim();
                    outBuf.AppendLine(SubVars(text));
                    continue;
                }
                if (up.StartsWith("A:"))
                {
                    var body = line.Substring(2);
                    var eq = body.IndexOf('=');
                    if (eq >= 0)
                    {
                        var name = body.Substring(0, eq).Trim().ToUpperInvariant();
                        var value = body.Substring(eq + 1).Trim();
                        _vars[name] = SubVars(value);
                    }
                    continue;
                }
                if (up.StartsWith("U:"))
                {
                    // U:NAME?PROMPT stubbed
                    var body = line.Substring(2);
                    var q = body.IndexOf('?');
                    if (q >= 0)
                    {
                        var name = body.Substring(0, q).Trim().ToUpperInvariant();
                        var prompt = body.Substring(q + 1).Trim();
                        _vars[name] = "(input)";
                        outBuf.AppendLine($"📝 {prompt} => (input)");
                    }
                    continue;
                }
                if (up.StartsWith("C:"))
                {
                    var body = line.Substring(2);
                    var parts = body.Split(';');
                    if (parts.Length == 2)
                    {
                        var pattern = SubVars(parts[0].Trim());
                        var text = SubVars(parts[1].Trim());
                        _lastCondition = pattern == text;
                        if (_lastCondition) outBuf.AppendLine(text);
                    }
                    continue;
                }
                if (up.StartsWith("M:"))
                {
                    var body = line.Substring(2);
                    var eq = body.IndexOf('=');
                    if (eq >= 0)
                    {
                        var left = SubVars(body.Substring(0, eq).Trim());
                        var right = SubVars(body.Substring(eq + 1).Trim());
                        _lastCondition = string.Equals(left, right, StringComparison.OrdinalIgnoreCase);
                    }
                    else
                    {
                        var parts2 = body.Split(';');
                        if (parts2.Length == 2)
                        {
                            _lastCondition = string.Equals(SubVars(parts2[0].Trim()), SubVars(parts2[1].Trim()), StringComparison.OrdinalIgnoreCase);
                        }
                        else _lastCondition = false;
                    }
                    continue;
                }
                if (up.StartsWith("Y:"))
                {
                    if (_lastCondition)
                    {
                        var inner = line.Substring(2).Trim();
                        var (embeddedOut, didJump) = ExecuteEmbedded(inner, ref pc);
                        outBuf.Append(embeddedOut);
                        if (didJump) continue;
                    }
                    continue;
                }
                if (up.StartsWith("N:"))
                {
                    if (!_lastCondition)
                    {
                        var inner = line.Substring(2).Trim();
                        var (embeddedOut, didJump) = ExecuteEmbedded(inner, ref pc);
                        outBuf.Append(embeddedOut);
                        if (didJump) continue;
                    }
                    continue;
                }
                if (up.StartsWith("J:"))
                {
                    var target = line.Substring(2).Trim().ToUpperInvariant();
                    if (_labels.TryGetValue(target, out var dest)) pc = dest; else outBuf.AppendLine($"❌ Unknown label: {target}");
                    continue;
                }
                if (up.StartsWith("E:"))
                {
                    var text = line.Substring(2).Trim();
                    outBuf.AppendLine($"❌ {SubVars(text)}");
                    continue;
                }
                if (up.StartsWith("Y:") || up.StartsWith("N:") || up.StartsWith("M:")) { continue; }
                outBuf.AppendLine($"ℹ️ Unknown: {line}");
            }
            if (steps >= 10000) outBuf.AppendLine("❌ Stopped: too many steps");
            return outBuf.ToString();
        }

        private string SubVars(string s)
        {
            foreach (var kv in _vars)
            {
                s = s.Replace($"{{{kv.Key}}}", kv.Value, StringComparison.OrdinalIgnoreCase);
                s = s.Replace($"${kv.Key}", kv.Value, StringComparison.OrdinalIgnoreCase);
            }
            return s;
        }

        private (string output, bool didJump) ExecuteEmbedded(string cmd, ref int pc)
        {
            var up = cmd.ToUpperInvariant();
            if (up.StartsWith("T:"))
            {
                var text = cmd.Substring(2).Trim();
                return (SubVars(text) + "\n", false);
            }
            if (up.StartsWith("E:"))
            {
                var text = cmd.Substring(2).Trim();
                return ($"❌ {SubVars(text)}\n", false);
            }
            if (up.StartsWith("A:"))
            {
                var body = cmd.Substring(2);
                var eq = body.IndexOf('=');
                if (eq >= 0)
                {
                    var name = body.Substring(0, eq).Trim().ToUpperInvariant();
                    var value = body.Substring(eq + 1).Trim();
                    _vars[name] = SubVars(value);
                }
                return (string.Empty, false);
            }
            if (up.StartsWith("J:"))
            {
                var target = cmd.Substring(2).Trim().ToUpperInvariant();
                if (_labels.TryGetValue(target, out var dest)) { pc = dest; return (string.Empty, true); }
                return ($"❌ Unknown label: {target}\n", false);
            }
            return ($"ℹ️ Unknown: {cmd}\n", false);
        }
    }
}

File: platforms/apple/TimeWarp/PilotExecutor.swift
import Foundation

class PilotExecutor {
    struct Label { let name: String; let index: Int }
    private var labels: [String: Int] = [:]
    private var vars: [String: String] = [:]
    private var lastCondition: Bool = false

    func execute(code: String) -> String {
        let rawLines = code.components(separatedBy: .newlines)
        var lines: [String] = []
        for l in rawLines { let t = l.trimmingCharacters(in: .whitespaces); if !t.isEmpty { lines.append(t) } }
        // First pass: collect labels L:NAME
        for (idx, line) in lines.enumerated() {
            if line.uppercased().hasPrefix("L:") {
                let name = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                if !name.isEmpty { labels[name.uppercased()] = idx }
            }
        }
    var pc = 0
        var output = ""
        var steps = 0
        let maxSteps = 10000

        func substituteVars(_ s: String) -> String {
            var result = s
            for (k,v) in vars { result = result.replacingOccurrences(of: "{\(k)}", with: v) }
            return result
        }

        while pc < lines.count && steps < maxSteps {
            steps += 1
            let line = lines[pc]
            pc += 1
            let up = line.uppercased()
            if up.hasPrefix("L:") { continue }
            if up.hasPrefix("R:") { break }
            if up.hasPrefix("T:") {
                let text = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                output += substituteVars(text) + "\n"
                continue
            }
            if up.hasPrefix("A:") {
                // A:NAME=VALUE
                let body = String(line.dropFirst(2))
                if let eq = body.firstIndex(of: "=") {
                    let name = body[..<eq].trimmingCharacters(in: .whitespaces).uppercased()
                    let value = body[body.index(after: eq)...].trimmingCharacters(in: .whitespaces)
                    vars[name] = substituteVars(value)
                }
                continue
            }
            if up.hasPrefix("U:") {
                // U:NAME?PROMPT  (store user input)
                let body = String(line.dropFirst(2))
                if let q = body.firstIndex(of: "?") {
                    let name = body[..<q].trimmingCharacters(in: .whitespaces).uppercased()
                    let prompt = body[body.index(after: q)...].trimmingCharacters(in: .whitespaces)
                    // No actual interactive input in this environment; stub.
                    vars[name] = "(input)"
                    output += "📝 \(prompt) => (input)\n"
                }
                continue
            }
            if up.hasPrefix("C:") {
                // C: pattern ; text-if-match
                let body = String(line.dropFirst(2))
                let parts = body.split(separator: ";", maxSplits: 1).map { String($0).trimmingCharacters(in: .whitespaces) }
                if parts.count == 2 {
                    let pattern = substituteVars(parts[0])
                    let text = substituteVars(parts[1])
                    lastCondition = (pattern == text)
                    if lastCondition { output += text + "\n" }
                }
                continue
            }
            if up.hasPrefix("M:") {
                // M: a = b   OR   M: a ; b
                let body = String(line.dropFirst(2))
                if let eq = body.firstIndex(of: "=") {
                    let left = substituteVars(String(body[..<eq]).trimmingCharacters(in: .whitespaces))
                    let right = substituteVars(String(body[body.index(after: eq)...]).trimmingCharacters(in: .whitespaces))
                    lastCondition = (left == right)
                } else {
                    let parts = body.split(separator: ";", maxSplits: 1).map { String($0).trimmingCharacters(in: .whitespaces) }
                    if parts.count == 2 {
                        lastCondition = (substituteVars(parts[0]) == substituteVars(parts[1]))
                    } else {
                        lastCondition = false
                    }
                }
                continue
            }
            if up.hasPrefix("Y:") {
                if lastCondition {
                    let inner = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                    // Execute a single embedded command like T:/J:/E:/A:
                    let res = executeEmbedded(inner, labels: labels, pc: &pc)
                    output += res.output
                    if res.didJump { continue }
                }
                continue
            }
            if up.hasPrefix("N:") {
                if !lastCondition {
                    let inner = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                    let res = executeEmbedded(inner, labels: labels, pc: &pc)
                    output += res.output
                    if res.didJump { continue }
                }
                continue
            }
            if up.hasPrefix("J:") {
                let target = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces).uppercased()
                if let dest = labels[target] { pc = dest } else { output += "❌ Unknown label: \(target)\n" }
                continue
            }
            if up.hasPrefix("E:") {
                let text = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                output += "❌ \(substituteVars(text))\n"
                continue
            }
            if up.hasPrefix("Y:") || up.hasPrefix("N:") || up.hasPrefix("M:") {
                // handled above
                continue
            }
            output += "ℹ️ Unknown: \(line)\n"
        }
        if steps >= maxSteps { output += "❌ Stopped: too many steps\n" }
        return output
    }

    private func executeEmbedded(_ cmd: String, labels: [String:Int], pc: inout Int) -> (output: String, didJump: Bool) {
        let up = cmd.uppercased()
        if up.hasPrefix("T:") {
            let text = String(cmd.dropFirst(2)).trimmingCharacters(in: .whitespaces)
            return (substituteVars(text) + "\n", false)
        }
        if up.hasPrefix("E:") {
            let text = String(cmd.dropFirst(2)).trimmingCharacters(in: .whitespaces)
            return ("❌ " + substituteVars(text) + "\n", false)
        }
        if up.hasPrefix("J:") {
            let target = String(cmd.dropFirst(2)).trimmingCharacters(in: .whitespaces).uppercased()
            if let dest = labels[target] { pc = dest; return ("", true) }
            return ("❌ Unknown label: \(target)\n", false)
        }
        if up.hasPrefix("A:") {
            let body = String(cmd.dropFirst(2))
            if let eq = body.firstIndex(of: "=") {
                let name = body[..<eq].trimmingCharacters(in: .whitespaces).uppercased()
                let value = body[body.index(after: eq)...].trimmingCharacters(in: .whitespaces)
                vars[name] = substituteVars(value)
            }
            return ("", false)
        }
        return ("ℹ️ Unknown: \(cmd)\n", false)
    }
}
